### PR TITLE
Step 11. Create Web Service

### PR DESCRIPTION
**Pull request by awilliamson1889**

### CODE REVIEW

File: department_app/rest/address_api.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+from department_app.models.app_models import Address
+from department_app.schemas.address_schema import AddressModel
+from department_app.models.app_models import db
+
+address_api = Blueprint('address_api', __name__)
+
+api = Api(address_api)
+
+
+class AddressInfo(Resource):
```
Comment (jimmyhollywood) [inline department_app/rest/address_api.py]: For API endpoint names let's use the next naming:
- for the single entity - just a name of class (AddressInfo -> Address)
- for the list of entities - postfix 'list' (AllAddressesInfo -> AddressList)
And the same for all endpoints

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -44,7 +44,7 @@ jobs:
           pip install pylint-flask-sqlalchemy
       - name: Run python mypy
         run:
-          python mypy.py
+          python init_test_data.py
```
Comment (jimmyhollywood) [inline .github/workflows/build.yml]: This is not so good way to initialize database with test data. You can use migration or fixtures

File: department_app/__init__.py [hunk 3]
```diff
@@ -44,8 +44,26 @@ def create_app(test_config=None):
...
+
+    import department_app.rest.skill_api
+    app.register_blueprint(department_app.rest.skill_api.skill_api)
+
+    import department_app.rest.address_api
+    app.register_blueprint(department_app.rest.address_api.address_api)
+
+    from department_app.models.app_models import Employee, Department, Permission, Address, Location, Skill
+
     return app
```
Comment (jimmyhollywood) [inline department_app/__init__.py]: As you're using app factory - move the code from __init__.py to the app.py file

File: department_app/rest/address_api.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            required: true
+            type: "integer"
+            format: "int64"
+        responses:
+          404:
+            description: Could not find address
+          200:
+            description: Address information returned
+        """
+        if not str(address_id).isdigit():
```
Comment (jimmyhollywood) [inline department_app/rest/address_api.py]: As I understand you have implemented validation using Pydantic, right? Can we now use these classes (pydantic) to validate requests?

File: department_app/tests/factoryes/department_factory.py [hunk 5]
```diff
@@ -8,4 +8,4 @@ class Meta:
```
Comment (jimmyhollywood) [inline department_app/tests/factoryes/department_factory.py:8]: factoryes -> factories

**Review by jimmyhollywood — CHANGES_REQUESTED** [review]
Generaly:
- Tests: as you have a lot of duplicated code (creation application) you can move it to the separate function.
- Tests directory doesn't mirror the structure of the source code. If you have views, service in src -> you should create the same structure in tests folder and put to these dirs corresponding test cases
- The next thing that you have missed - is running your application with gunicorn
Let's update all changes - if you any questions please contact me.

File: config.py [hunk 6]
```diff
@@ -26,3 +26,8 @@ class TestingConfig(Config):
     # SQLALCHEMY_DATABASE_URI = "postgresql://postgres:postgres@localhost:5432/flask_test"
     SQLALCHEMY_DATABASE_URI = 'sqlite:///flask_app.db'
     TESTING = True
+
+
+configs = {'Prod': ProductionConfig,
```
Comment (jimmyhollywood) [inline config.py]: This is not needed here. When you run app using gunicorn just use environment variable with name of config class (Development, Production, Testing)

File: department_app/app.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
+"""web app"""
+import os
+from flask import Flask
+from flask_migrate import Migrate
+from flasgger import Swagger
+from department_app.models import db
+
+
+MIGRATION_DIR = os.path.join('department_app/migrations')
```
Comment (jimmyhollywood) [inline department_app/app.py]: Move to config

File: department_app/app.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def create_app(config):
+    """Create app method"""
+
+    app = Flask(__name__)
+    app.config.from_object("config." + config)
```
Comment (jimmyhollywood) [inline department_app/app.py]: Just use env variable in appropriate format (and add note in readme):
```export APP_SETTINGS=config.DevelopmentConfig```

File: department_app/app.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    from department_app.rest.permission import permission_api
+    app.register_blueprint(permission_api)
+
+    from department_app.rest.skill import skill_api
+    app.register_blueprint(skill_api)
+
+    from department_app.rest.address import address_api
+    app.register_blueprint(address_api)
+
+    from department_app.models import EmployeeModel, DepartmentModel, PermissionModel,\
```
Comment (jimmyhollywood) [inline department_app/app.py]: Check these imports. They might be avoided

File: department_app/app.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        "static_url_path": "/flasgger_static",
+        "swagger_ui": True,
+        "specs_route": "/swagger/",
+    }
+
+    swagger = Swagger(app, config=swagger_config)
+
+    migrate = Migrate(app, db, directory=MIGRATION_DIR)
+
+    from department_app.views.routes import frontend
```
Comment (jimmyhollywood) [inline department_app/app.py]: Let's put all import one by one in the same place

File: department_app/fixtures/address.yaml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+- table: Address
+  records:
+  - id: 1
```
Comment (jimmyhollywood) [inline department_app/fixtures/address.yaml]: Do you really need to define the ids? The same for others yaml files

File: department_app/models/__init__.py [hunk 12]
```diff
@@ -0,0 +1,10 @@
+from flask_sqlalchemy import SQLAlchemy
```
Comment (jimmyhollywood) [inline department_app/models/__init__.py]: You should not import in the init you real dependencies. Let's move it to the separate file (db.py or database.py in the root)

File: department_app/models/__init__.py [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+
+from .address import Address as AddressModel
+from .department import Department as DepartmentModel
+from .employee import Employee as EmployeeModel
+from .location import Location as LocationModel
+from .permission import Permission as PermissionModel
+from .skill import Skill as SkillModel
+
+db = SQLAlchemy()
```
Comment (jimmyhollywood) [inline department_app/models/__init__.py]: The same

File: department_app/models/address.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
+"""Model for Address"""
+from dataclasses import dataclass
+from flask_sqlalchemy import SQLAlchemy
+db = SQLAlchemy()
```
Comment (jimmyhollywood) [inline department_app/models/address.py]: This line is duplicated in each model module. This thing should be imported from the single file where it is defined (db.py or database.py)

File: department_app/rest/address.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
+"""Rest addresses Api"""
+from flask import Blueprint, request, jsonify, make_response
+from flask_restful import Resource, Api, abort
+from pydantic import ValidationError
+from department_app.models import AddressModel, db
```
Comment (jimmyhollywood) [inline department_app/rest/address.py]: models should not be used in views

File: department_app/rest/address.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+        try:
+            address = AddressSchema(**address_data)
+        except ValidationError as exception:
+            abort(404, message=f"Exception: {exception}")
+
+        result = AddressModel(**address_data)
+
+        db.session.add(result)
+        db.session.commit()
+        return address.dict(), 201
```
Comment (jimmyhollywood) [inline department_app/rest/address.py]: Let's use the same response creation approach (like you've used above with make_response)

File: department_app/service/address.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+from department_app.models import db, AddressModel
+from flask_restful import abort
+
+
+class CRUDAddress:
+    """Address CRUD class"""
+    @staticmethod
+    def get_address(address_id):
+        """Get all address func"""
+        if not str(address_id).isdigit():
```
Comment (jimmyhollywood) [inline department_app/service/address.py]: Format validation is implemented in view level and in pydantic. So, here you suppose that data is valid.

File: department_app/wsgi.py [hunk 18]
```diff
@@ -1,4 +1,6 @@
 """WSGI file"""
-from department_app import create_app
+from department_app.app import create_app
+import os
 
-app = create_app('Prod')
+os.environ['FLASK_CONFIG'] = 'TestingConfig'
```
Comment (jimmyhollywood) [inline department_app/wsgi.py]: This should not be here. Instead use export (on linux) command.

File: .github/workflows/build.yml [hunk 19]
```diff
@@ -36,15 +36,9 @@ jobs:
...
-          pip install coverage
-          pip install flask-migrate --upgrade
           pip install -r requirements.txt
-          pip install pylint
-          pip install pylint-flask
-          pip install pylint-flask-sqlalchemy
-      - name: Run python mypy
-        run:
-          python mypy.py
+          pip install -r CI-requirements.txt
```
Comment (jimmyhollywood) [inline .github/workflows/build.yml:39]: requirements.txt
```
flask==1.1.1
pydantic==1.1.1
...
```
test_requirements.txt
```
-r requirements.txt
pytest==1.1.1
```

File: .pylintrc [hunk 20]
```diff
@@ -13,6 +13,7 @@ fail-under=10.0
 ignore=CVS,
       migrations,
       manage.py,
+      config.py,
       asgi.py,
       mypy.py,
```
Comment (jimmyhollywood) [inline .pylintrc]: mypy still exist here?

File: README.md [hunk 21]
```diff
@@ -2,28 +2,48 @@
...
+$ exit
+```
 
 # Create and run migrations
 
 
 ```{bash}
+$ export FLASK_APP=department_app.app
 $ flask db init
 $ flask db migrate -m "Initial migration."
```
Comment (jimmyhollywood) [inline README.md:25]: Migration files should be presented in sources. User should not generate these files himself

File: department_app/service/address.py [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class CRUDAddress:
+    """Address CRUD class"""
+    @staticmethod
+    def get(address_id: int):
+        """Get address func"""
+        try:
+            address = AddressModel.query.filter_by(id=address_id).first()
+        except DataError as exception:
```
Comment (jimmyhollywood) [inline department_app/service/address.py]: Each exception should be handled. In this case you should store an error message in log file. And then - reraise exception

File: department_app/service/address.py [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+            raise exception
+
+        if not address:
+            return None
+        return address
+
+    @staticmethod
+    def update(address_id: int, name: str):
+        """update address func"""
+        address = CRUDAddress.get(address_id)
```
Comment (jimmyhollywood) [inline department_app/service/address.py]: Let's use the following approach:
```
try:
    # We don't need to check whether record is existing - we just trying to update it
    address_model.query.update(name=name).where(id=id) # i am no sure about syntax
except IntegrityError:
    log.write(msg="Address with id %d does not exist")
    raise # or return False
return True
```

**Review by jimmyhollywood — CHANGES_REQUESTED** [review]

File: config.py (hunk 6 above)
Comment (awilliamson1889) [reply config.py]: Unnecessary variable deleted.

File: department_app/app.py (hunk 9 above)
Comment (awilliamson1889) [reply department_app/app.py]: These imports already deleted.

File: department_app/app.py (hunk 10 above)
Comment (awilliamson1889) [reply department_app/app.py]: The imports location has been changed.

File: .github/workflows/build.yml (hunk 19 above)
Comment (awilliamson1889) [reply .github/workflows/build.yml:39]: "CI-requirements.txt" updated.

File: department_app/fixtures/address.yaml (hunk 11 above)
Comment (awilliamson1889) [reply department_app/fixtures/address.yaml]: "id" fields deleted.

File: department_app/models/__init__.py (hunk 12 above)
Comment (awilliamson1889) [reply department_app/models/__init__.py]: This import already deleted.

File: department_app/models/__init__.py (hunk 13 above)
Comment (awilliamson1889) [reply department_app/models/__init__.py]: This import already deleted.

File: department_app/rest/address.py (hunk 15 above)
Comment (awilliamson1889) [reply department_app/rest/address.py]: This import already deleted.

File: department_app/rest/address.py (hunk 16 above)
Comment (awilliamson1889) [reply department_app/rest/address.py]: All implemented methods have same return format..

File: department_app/service/address.py (hunk 22 above)
Comment (awilliamson1889) [reply department_app/service/address.py]: Now error message storing in "app.log" file.

File: department_app/service/address.py (hunk 23 above)
Comment (awilliamson1889) [reply department_app/service/address.py]: This method not check address existence, address check existence happens in REST module.

Now error message storing in "app.log" file.

File: .pylintrc (hunk 20 above)
Comment (awilliamson1889) [reply .pylintrc]: Deleted.

File: README.md (hunk 21 above)
Comment (awilliamson1889) [reply README.md:25]: Now migration files presented in source.

File: department_app/app.py (hunk 8 above)
Comment (awilliamson1889) [reply department_app/app.py]: Now create_app method use ENV variable.

File: department_app/app.py (hunk 7 above)
Comment (awilliamson1889) [reply department_app/app.py]: "MIGRATION_DIR" moved to config.

File: department_app/wsgi.py (hunk 18 above)
Comment (awilliamson1889) [reply department_app/wsgi.py]: Done.

File: .github/workflows/build.yml (hunk 19 above)
Comment (jimmyhollywood) [reply .github/workflows/build.yml:39]: So then it's enough to install only CI-requirements.txt

File: .github/workflows/build.yml [hunk 24]
```diff
@@ -36,20 +36,18 @@ jobs:
...
+
       - name: Run pylint check
         run:
           pylint --load-plugins pylint_flask_sqlalchemy department_app
+      - name: export APP_SETTINGS
+        run:
+          export APP_SETTINGS=department_app.config.TestingConfig
```
Comment (jimmyhollywood) [inline .github/workflows/build.yml:46]: I definitely don't like this prefix `department_app`

File: README.md [hunk 25]
```diff
@@ -2,28 +2,47 @@
...
+postgres=# CREATE DATABASE gallery;
+postgres=# GRANT ALL PRIVILEGES ON DATABASE gallery to jacky;
+postgres=# \q
+$ exit
+```
 
-# Create and run migrations
+# Create ENV variables
+```{bash}
+export APP_SETTINGS=department_app.config.ProductionConfig
```
Comment (jimmyhollywood) [inline README.md:31]: The same - prefix

File: department_app/rest/address.py [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+            format: "int64"
+        responses:
+          404:
+            description: Could not find address
+          200:
+            description: Address information returned
+        """
+        address = Address.get_address(address_id)
+
+        Address.check_exist(address_id, address)
```
Comment (jimmyhollywood) [inline department_app/rest/address.py]: Which kind of check is performed here?

File: department_app/rest/address.py [hunk 27]
```diff
@@ -0,0 +1,178 @@
...
+
+        address_json = Address.get_json()
+
+        Address.validate_json(address_json)
+
+        try:
+            result = CRUDAddress.update(address_id, name=address_json['name'])
+        except IntegrityError as exception:
+            abort(404, message=f"{exception}")
+        return make_response(jsonify(result), 201)
```
Comment (jimmyhollywood) [inline department_app/rest/address.py]: Again, I see business logic in this view.

**Review by jimmyhollywood — CHANGES_REQUESTED** [review]

File: department_app/service/permission.py [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+    @staticmethod
+    def update(permission_id, name):
+        """Update permission func"""
+        logging.info("Update permission method called with parameters: id=%s, name=%s.", permission_id, name)
+
+        try:
+            result = PermissionModel.query.where(PermissionModel.id == permission_id). \
+                update({PermissionModel.name: name})
+            db.session.commit()
+        except IntegrityError as exception:
```
Comment (jimmyhollywood) [inline department_app/service/permission.py]: Just to note: you can get rid of `as exception` and just use `raise`. Looks like:
```
try:
    # bad code
except IntegrityError:
    # doing smth useful
    raise # will just raise the latest exception
```

**Review by jimmyhollywood — APPROVED** [review]

**Comment by jimmyhollywood** [comment]
I see failed check in action pipeline and unresolved conflicts. So, please check your pipeline (maybe it makes sense to define some threshold value for coverage) and resolve conflicting files before merging.

**Pull request merged**